File: faza3/Blockchain.java
// Meno študenta:
// AI: Implementácia vytvorená s pomocou Claude AI (Anthropic) - kompletná implementácia
// Blockchain triedy s podporou forkov (stromová štruktúra), per-block UTXOPool,
// globálny TransactionPool a CUT_OFF_AGE limit.

// Blockchain by mal na naplnenie funkcií udržiavať iba obmedzené množstvo uzlov
// Nemali by ste mať všetky bloky pridané do blockchainu v pamäti
// pretože by to mohlo spôsobiť pretečenie pamäte.
import java.util.ArrayList;
import java.util.HashMap;

public class Blockchain {
    public static final int CUT_OFF_AGE = 12;

    // Všetky potrebné informácie na spracovanie bloku v reťazi blokov
    private class BlockNode {
        public Block b;
        public BlockNode parent;
        public ArrayList<BlockNode> children;
        public int height;
        // utxo pool na vytvorenie nového bloku na vrchu tohto bloku
        private UTXOPool uPool;

        public BlockNode(Block b, BlockNode parent, UTXOPool uPool) {
            this.b = b;
            this.parent = parent;
            children = new ArrayList<BlockNode>();
            this.uPool = uPool;
            if (parent != null) {
                height = parent.height + 1;
                parent.children.add(this);
            } else {
                height = 1;
            }
        }

        public UTXOPool getUTXOPoolCopy() {
            return new UTXOPool(uPool);
        }
    }

    // Mapovanie hash bloku na BlockNode pre rýchle vyhľadávanie
    private HashMap<ByteArrayWrapper, BlockNode> blockMap;
    // Globálny pool transakcií
    private TransactionPool txPool;
    // Najvyšší blok (najstarší ak je viac v rovnakej výške)
    private BlockNode maxHeightNode;

    /**
     * Vytvor prázdny blockchain iba s prvým (Genesis) blokom. Predpokladajme, že
     * {@code genesisBlock} je platný blok.
     */
    public Blockchain(Block genesisBlock) {
        blockMap = new HashMap<ByteArrayWrapper, BlockNode>();
        txPool = new TransactionPool();

        // Vytvorenie UTXOPool pre genesis blok - pridanie coinbase výstupu
        UTXOPool genesisPool = new UTXOPool();
        Transaction coinbase = genesisBlock.getCoinbase();
        for (int i = 0; i < coinbase.numOutputs(); i++) {
            UTXO utxo = new UTXO(coinbase.getHash(), i);
            genesisPool.addUTXO(utxo, coinbase.getOutput(i));
        }

        BlockNode genesisNode = new BlockNode(genesisBlock, null, genesisPool);
        ByteArrayWrapper genesisHash = new ByteArrayWrapper(genesisBlock.getHash());
        blockMap.put(genesisHash, genesisNode);
        maxHeightNode = genesisNode;
    }

    /** Získaj najvyšší (maximum height) blok */
    public Block getBlockAtMaxHeight() {
        return maxHeightNode.b;
    }

    /**
     * Získaj UTXOPool na ťaženie nového bloku na vrchu najvyššieho (max height) bloku
     */
    public UTXOPool getUTXOPoolAtMaxHeight() {
        return maxHeightNode.getUTXOPoolCopy();
    }

    /** Získaj pool transakcií na vyťaženie nového bloku */
    public TransactionPool getTransactionPool() {
        return txPool;
    }

    /**
     * Pridaj {@code block} do blockchainu, ak je platný. Kvôli platnosti by mali
     * byť všetky transakcie platné a blok by mal byť na
     * {@code height > (maxHeight - CUT_OFF_AGE)}.
     *
     * @return true, ak je blok úspešne pridaný
     */
    public boolean blockAdd(Block block) {
        // Odmietni genesis bloky (prevBlockHash je null)
        if (block.getPrevBlockHash() == null) {
            return false;
        }

        // Nájdi rodičovský blok
        ByteArrayWrapper prevHash = new ByteArrayWrapper(block.getPrevBlockHash());
        BlockNode parentNode = blockMap.get(prevHash);

        // Ak rodičovský blok neexistuje, odmietni
        if (parentNode == null) {
            return false;
        }

        // Skontroluj, či nový blok by bol na height > (maxHeight - CUT_OFF_AGE)
        int newHeight = parentNode.height + 1;
        if (newHeight <= maxHeightNode.height - CUT_OFF_AGE) {
            return false;
        }

        // Vytvor UTXOPool pre nový blok na základe rodičovského UTXOPool
        UTXOPool parentPool = parentNode.getUTXOPoolCopy();

        // Pridaj coinbase UTXO nového bloku do poolu
        Transaction coinbase = block.getCoinbase();
        for (int i = 0; i < coinbase.numOutputs(); i++) {
            UTXO utxo = new UTXO(coinbase.getHash(), i);
            parentPool.addUTXO(utxo, coinbase.getOutput(i));
        }

        // Skontroluj platnosť všetkých transakcií v bloku
        HandleTxs handler = new HandleTxs(parentPool);
        Transaction[] blockTxs = block.getTransactions().toArray(new Transaction[0]);
        Transaction[] validTxs = handler.handler(blockTxs);

        // Všetky transakcie v bloku musia byť platné
        if (validTxs.length != blockTxs.length) {
            return false;
        }

        // Získaj aktualizovaný UTXOPool po spracovaní transakcií
        UTXOPool newPool = handler.UTXOPoolGet();

        // Vytvor nový BlockNode
        BlockNode newNode = new BlockNode(block, parentNode, newPool);
        ByteArrayWrapper blockHash = new ByteArrayWrapper(block.getHash());
        blockMap.put(blockHash, newNode);

        // Aktualizuj maxHeightNode ak je nový blok vyšší
        if (newHeight > maxHeightNode.height) {
            maxHeightNode = newNode;
        }

        // Odstráň transakcie z poolu, ktoré boli zahrnuté do bloku
        for (Transaction tx : block.getTransactions()) {
            txPool.removeTransaction(tx.getHash());
        }

        // Odstráň staré bloky, ktoré sú príliš hlboko (pod CUT_OFF_AGE)
        // Pre jednoduchosť to robíme tak, že odstraňujeme bloky z mapy
        ArrayList<ByteArrayWrapper> toRemove = new ArrayList<ByteArrayWrapper>();
        for (HashMap.Entry<ByteArrayWrapper, BlockNode> entry : blockMap.entrySet()) {
            if (entry.getValue().height < maxHeightNode.height - CUT_OFF_AGE) {
                toRemove.add(entry.getKey());
            }
        }
        for (ByteArrayWrapper key : toRemove) {
            blockMap.remove(key);
        }

        return true;
    }

    /** Pridaj transakciu do transakčného poolu */
    public void transactionAdd(Transaction tx) {
        txPool.addTransaction(tx);
    }
}

File: python/blockchain.py
# AI: Implementacia vytvorena s pomocou Claude AI (Anthropic) - Python rewrite (bonus 5 bodov).
# Blockchain - stromova struktura s forkmi (Faza 3).

from utxo import UTXO, UTXOPool
from transaction_pool import TransactionPool
from handle_txs import HandleTxs


CUT_OFF_AGE = 12


class BlockNode:
    """Uzol v strome blockchainu."""

    def __init__(self, block, parent, utxo_pool):
        self.block = block
        self.parent = parent
        self.children = []
        self.utxo_pool = utxo_pool

        if parent is not None:
            self.height = parent.height + 1
            parent.children.append(self)
        else:
            self.height = 1

    def get_utxo_pool_copy(self):
        return UTXOPool(self.utxo_pool)


class Blockchain:
    """Blockchain s podporou forkov a CUT_OFF_AGE."""

    def __init__(self, genesis_block):
        self.block_map = {}
        self.tx_pool = TransactionPool()

        # UTXOPool pre genesis blok
        genesis_pool = UTXOPool()
        coinbase = genesis_block.get_coinbase()
        for i in range(coinbase.num_outputs()):
            utxo = UTXO(coinbase.get_hash(), i)
            genesis_pool.add_utxo(utxo, coinbase.get_output(i))

        genesis_node = BlockNode(genesis_block, None, genesis_pool)
        genesis_hash = genesis_block.get_hash()
        self.block_map[genesis_hash] = genesis_node
        self.max_height_node = genesis_node

    def get_block_at_max_height(self):
        return self.max_height_node.block

    def get_utxo_pool_at_max_height(self):
        return self.max_height_node.get_utxo_pool_copy()

    def get_transaction_pool(self):
        return self.tx_pool

    def block_add(self, block):
        """Prida blok do blockchainu ak je platny."""
        # Odmietni genesis bloky
        if block.get_prev_block_hash() is None:
            return False

        # Najdi rodicovsky blok
        prev_hash = block.get_prev_block_hash()
        parent_node = self.block_map.get(prev_hash)
        if parent_node is None:
            return False

        # Skontroluj vysku
        new_height = parent_node.height + 1
        if new_height <= self.max_height_node.height - CUT_OFF_AGE:
            return False

        # Vytvor UTXOPool z rodica
        parent_pool = parent_node.get_utxo_pool_copy()

        # Pridaj coinbase
        coinbase = block.get_coinbase()
        for i in range(coinbase.num_outputs()):
            utxo = UTXO(coinbase.get_hash(), i)
            parent_pool.add_utxo(utxo, coinbase.get_output(i))

        # Validuj transakcie
        handler = HandleTxs(parent_pool)
        block_txs = block.get_transactions()
        valid_txs = handler.handler(block_txs)

        if len(valid_txs) != len(block_txs):
            return False

        # Pridaj novy uzol
        new_pool = handler.utxo_pool_get()
        new_node = BlockNode(block, parent_node, new_pool)
        self.block_map[block.get_hash()] = new_node

        # Aktualizuj max height
        if new_height > self.max_height_node.height:
            self.max_height_node = new_node

        # Odstran transakcie z poolu
        for tx in block.get_transactions():
            self.tx_pool.remove_transaction(tx.get_hash())

        # Cleanup starych blokov
        to_remove = [h for h, node in self.block_map.items()
                     if node.height < self.max_height_node.height - CUT_OFF_AGE]
        for h in to_remove:
            del self.block_map[h]

        return True

    def transaction_add(self, tx):
        self.tx_pool.add_transaction(tx)
